test/test_cuda_nvml_based_avail.py

```diff
@@ -0,0 +1,71 @@
+# Owner(s): ["module: cuda"]
+
+import os
+import sys
+import multiprocessing
+import torch
+import unittest
+from unittest.mock import patch
+
+# NOTE: Each of the tests in this module need to be run in a brand new process to ensure CUDA is uninitialized
+# prior to test initiation.
+# Before executing the desired tests, we need to disable CUDA initialization and fork_handler additions that would
+# otherwise be triggered by the `torch.testing._internal.common_utils` module import
+os.environ['PYTORCH_NVML_BASED_CUDA_CHECK'] = '1'
+from torch.testing._internal.common_utils import (parametrize, instantiate_parametrized_tests, run_tests, TestCase,
+                                                  IS_WINDOWS)
+# NOTE: Because `remove_device_and_dtype_suffixes` initializes CUDA context (triggered via the import of
+# `torch.testing._internal.common_device_type` which imports `torch.testing._internal.common_cuda`) we need
+# to bypass that method here which should be irrelevant to the parameterized tests in this module.
+torch.testing._internal.common_utils.remove_device_and_dtype_suffixes = lambda x: x
+
+TEST_CUDA = torch.cuda.is_available()
+if not TEST_CUDA:
+    print('CUDA not available, skipping tests', file=sys.stderr)
+    TestCase = object  # type: ignore[misc, assignment] # noqa: F811
+
+
+class TestExtendedCUDAIsAvail(TestCase):
+    SUBPROCESS_REMINDER_MSG = (
+        "\n REMINDER: Tests defined in test_cuda_nvml_based_avail.py must be run in a process "
+        "where there CUDA Driver API has not been initialized. Before further debugging, ensure you are either using "
+        "run_test.py or have added --subprocess to run each test in a different subprocess.")
+
+    def setUp(self):
+        super().setUp()
+        torch.cuda.device_count.cache_clear()  # clear the lru_cache on this method before our test
+        if os.getenv('PYTORCH_NVML_BASED_CUDA_CHECK'):
+            del os.environ['PYTORCH_NVML_BASED_CUDA_CHECK']
```

Comment:
should you restore os env in the teardown?

Reply:
I've now moved the os env config that was required for some of the test module imports behind a context guard so all os env modifications in this module (in the test case and outside of it) should be context guarded obviating the need for a teardown override.

Reply:
Thanks again for the review and your help @ngimel! :rocket: :tada: